packages/next/src/client/components/router-reducer/handle-mutable.ts

```diff
@@ -15,26 +19,28 @@ export function handleMutable(
   return {
     buildId: state.buildId,
     // Set href.
-    canonicalUrl:
-      mutable.canonicalUrl != null
-        ? mutable.canonicalUrl === state.canonicalUrl
-          ? state.canonicalUrl
-          : mutable.canonicalUrl
-        : state.canonicalUrl,
+    canonicalUrl: isNotUndefined(mutable.canonicalUrl)
```

Comment:
why don't write all of the logic in that file as
```
canonicalUrl: mutable?.canoniclaUrl ?? state.canonicalUrl,
```

Reply:
In React we just write !== undefined. (Don’t need to do the typeof thing in strict mode). It’s faster than calling a helper function because you don’t have an extra stack frame.

A function might compress better in some cases (although not always) but the way to think code size is that it’s mostly a heuristic for how long it takes for the VM to parse and compile (that’s why byte for byte JS is way more expensive than, say, an image); in this case the inlined form is very simple.

Reply:
What I was referring to is that optional chaining and `??` are downleveled, which causes them to explode in size as there are intermediate variables created. In this case we can probably just use `mutable.canonicalUrl !== undefined` though?